src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs

```diff
@@ -177,12 +179,21 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
                 {
                     Error(diagnostics, ErrorCode.ERR_AttributeCtorInParameter, node, attributeConstructor.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat));
                 }
+
+                boundConstructorArguments = BuildArgumentsForErrorRecovery(constructorArguments, ImmutableArray.Create(attributeConstructor));
```

Comment:
>BuildArgumentsForErrorRecovery [](start = 44, length = 30)

It looks like we are calling this non-discriminatively both for error and success cases. This doesn't feel right. It is quite possible I misunderstood what issue we are trying to deal with from the very beginning.

Reply:
The case we are dealing with is that, in error cases, there are some arguments that will need to be converted. I can make this conditional on whether there was a lookup failure though.

---
In reply to: [577675907](https://github.com/dotnet/roslyn/pull/51206#discussion_r577675907) [](ancestors = 577675907)

Reply:
> The case we are dealing with is that, in error cases, there are some arguments that will need to be converted. I can make this conditional on whether there was a lookup failure though.

This logic still feels somewhat suspicious to me. Where does conversion happen when lookup succeeds?

---
In reply to: [579374920](https://github.com/dotnet/roslyn/pull/51206#discussion_r579374920) [](ancestors = 579374920,577675907)

Reply:
> Where does conversion happen when lookup succeeds?

As part of `BindAttributeConstructor` above. The nested call to `TryPerformConstructorOverloadResolution` coerces the arguments if overload resolution succeeded (and for cases when an inaccessible constructor was found as well).

---
In reply to: [581140844](https://github.com/dotnet/roslyn/pull/51206#discussion_r581140844) [](ancestors = 581140844,579374920,577675907)

Reply:
> As part of BindAttributeConstructor above. The nested call to TryPerformConstructorOverloadResolution coerces the arguments if overload resolution succeeded (and for cases when an inaccessible constructor was found as well).

Then it feels to me that this operation should be performed based on whether ```TryPerformConstructorOverloadResolution``` succeeded and based on ```candidateConstructors``` that it produces, rather than based on ```resultKind```. Relying on ```resultKind``` feels fragile, exact correlation between its value and the failure ```TryPerformConstructorOverloadResolution``` is not obvious. Also, I think that ```SelectAsArrayInPlaceAndFree``` below should be called only when ```BindAttributeConstructor``` was not called.

---
In reply to: [582325554](https://github.com/dotnet/roslyn/pull/51206#discussion_r582325554) [](ancestors = 582325554,581140844,579374920,577675907)

Reply:
Sure, that seems like a reasonable change. I'll do that.

---
In reply to: [582350196](https://github.com/dotnet/roslyn/pull/51206#discussion_r582350196) [](ancestors = 582350196,582325554,581140844,579374920,577675907)